## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 514657`
Commit: [`061b4c6412714ea654a193bd42c82f6afef2620e`](https://github.com/NixOS/nixpkgs/commit/061b4c6412714ea654a193bd42c82f6afef2620e) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/061b4c6412714ea654a193bd42c82f6afef2620e..pull/514657/head))
Merge: [`a7566e25440fff386699fd7e567e71b3f4fcd2eb`](https://github.com/NixOS/nixpkgs/commit/a7566e25440fff386699fd7e567e71b3f4fcd2eb)

Logs: https://github.com/kilyanni/nixpkgs-review-gha/actions/runs/25186913665


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python313Packages.bittensor-wallet</li>
    <li>python313Packages.bittensor-wallet.dist</li>
    <li>python314Packages.bittensor-wallet</li>
    <li>python314Packages.bittensor-wallet.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python313Packages.bittensor-wallet</li>
    <li>python313Packages.bittensor-wallet.dist</li>
    <li>python314Packages.bittensor-wallet</li>
    <li>python314Packages.bittensor-wallet.dist</li>
  </ul>
</details>

---
### `x86_64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python314Packages.bittensor-wallet</li>
    <li>python314Packages.bittensor-wallet.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python313Packages.bittensor-wallet</li>
    <li>python313Packages.bittensor-wallet.dist</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>python314Packages.bittensor-wallet</summary>
<pre>        w = Wallet(name=wallet_name, hotkey=wallet_hotkey, path=wallet_path)
&gt;       w.create_hotkey_from_uri(&quot;//test&quot;, use_password=False, overwrite=True)
E       KeyFileError: Generic error: Failed to create hotkey from uri: FileOpen(&quot;Failed to open file: Permission denied (os error 13).&quot;)

tests/test_wallet.py:237: KeyFileError
=============================== warnings summary ===============================
../../../store/0pcnw68yhmp3l1kx2kiv6880v8jyf1qr-python3.14-ansible-vault-rw-2.1.0/lib/python3.14/site-packages/ansible_vault/parsing.py:20
  /nix/store/0pcnw68yhmp3l1kx2kiv6880v8jyf1qr-python3.14-ansible-vault-rw-2.1.0/lib/python3.14/site-packages/ansible_vault/parsing.py:20: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools&lt;81.
    from pkg_resources import parse_version

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_wallet.py::test_create_coldkey_from_uri - KeyFileError: Generic error: Failed to create coldkey from uri: Generic(&quot;Fa...
FAILED tests/test_wallet.py::test_hotkey_coldkey_from_uri - KeyFileError: Generic error: Failed to create hotkey from uri: FileOpen(&quot;Fa...
ERROR tests/test_keypair.py::test_keypair_type - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
ERROR tests/test_wallet.py::test_unlock_hotkey - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
ERROR tests/test_wallet.py::test_unlock_coldkey - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
ERROR tests/test_wallet.py::test_unlock_coldkeypub - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
ERROR tests/test_wallet.py::test_unlock_hotkeypub - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
======== 2 failed, 44 passed, 1 warning, 5 errors in 140.76s (0:02:20) =========</pre>
</details>
</details>

---
### `aarch64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python313Packages.bittensor-wallet</li>
    <li>python313Packages.bittensor-wallet.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python314Packages.bittensor-wallet</li>
    <li>python314Packages.bittensor-wallet.dist</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>python313Packages.bittensor-wallet</summary>
<pre>        w = Wallet(name=wallet_name, hotkey=wallet_hotkey, path=wallet_path)
&gt;       w.create_hotkey_from_uri(&quot;//test&quot;, use_password=False, overwrite=True)
E       KeyFileError: Generic error: Failed to create hotkey from uri: FileOpen(&quot;Failed to open file: Permission denied (os error 13).&quot;)

tests/test_wallet.py:237: KeyFileError
=============================== warnings summary ===============================
../../../store/giwaklf5gg1hqrqlzh3a2db4qbpfx6pz-python3.13-ansible-vault-rw-2.1.0/lib/python3.13/site-packages/ansible_vault/parsing.py:20
  /nix/store/giwaklf5gg1hqrqlzh3a2db4qbpfx6pz-python3.13-ansible-vault-rw-2.1.0/lib/python3.13/site-packages/ansible_vault/parsing.py:20: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools&lt;81.
    from pkg_resources import parse_version

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_wallet.py::test_create_coldkey_from_uri - KeyFileError: Generic error: Failed to create coldkey from uri: Generic(&quot;Fa...
FAILED tests/test_wallet.py::test_hotkey_coldkey_from_uri - KeyFileError: Generic error: Failed to create hotkey from uri: FileOpen(&quot;Fa...
ERROR tests/test_keypair.py::test_keypair_type - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
ERROR tests/test_wallet.py::test_unlock_hotkey - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
ERROR tests/test_wallet.py::test_unlock_coldkey - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
ERROR tests/test_wallet.py::test_unlock_coldkeypub - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
ERROR tests/test_wallet.py::test_unlock_hotkeypub - KeyFileError: Generic error: Failed to create new coldkey: KeyFileError(Gen...
============== 2 failed, 44 passed, 1 warning, 5 errors in 48.82s ==============</pre>
</details>
</details>

--- a/pkgs/development/python-modules/bittensor-wallet/default.nix
+++ b/pkgs/development/python-modules/bittensor-wallet/default.nix
@@ -1,0 +1,80 @@
+{
+  lib,
+  stdenv,
+  buildPythonPackage,
+  fetchFromGitHub,
+  rustPlatform,
+  pkg-config,
+  openssl,
+  libsodium,
+  libiconv,
+  pytestCheckHook,
+  ansible-vault-rw,
+  setuptools,
+}:
+
+buildPythonPackage (finalAttrs: {
+  pname = "bittensor-wallet";
+  version = "4.0.1";
+  pyproject = true;
+
+  __structuredAttrs = true;
+
+  src = fetchFromGitHub {
+    owner = "latent-to";
+    repo = "btwallet";
+    tag = "v${finalAttrs.version}";
+    hash = "sha256-L774RPoasixvW+0Z4WuJ6eLuazLQscckRU++VCAiFug=";
+  };
+
+  cargoDeps = rustPlatform.fetchCargoVendor {
+    inherit (finalAttrs) pname version src;
+    hash = "sha256-ETr7XhSmUTqtWDGzJMq5ijaLL8+tqmLJa/ngmzwWiFg=";
+  };
+
+  nativeBuildInputs = [
+    pkg-config
+  ]
+  ++ (with rustPlatform; [
+    cargoSetupHook
+    maturinBuildHook
+  ]);
+
+  buildInputs = [
+    openssl
+    libsodium
+  ]
+  ++ lib.optionals stdenv.hostPlatform.isDarwin [ libiconv ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    ansible-vault-rw
+    setuptools # ansible_vault uses pkg_resources at import time
+  ];
+
+  # tests hardcode /tmp/tests_wallets; /tmp is shared across builds on Darwin
+  postPatch = lib.optionalString stdenv.hostPlatform.isDarwin ''
+    walletTestPath=$(mktemp -d)
+    sed -i "s|/tmp/tests_wallets|$walletTestPath/tests_wallets|g" tests/test_wallet.py tests/test_keypair.py
+  '';
+
+  preCheck = ''
+    # remove source package so tests import the installed Rust extension
+    rm -rf bittensor_wallet
+    # ansible-vault writes to HOME
+    export HOME=$(mktemp -d)
+  '';
+
+  # test_common_calls.py requires bittensor and substrateinterface, which are not in nixpkgs.
+  disabledTestPaths = [ "tests/test_common_calls.py" ];
+
+  pythonImportsCheck = [ "bittensor_wallet" ];
+
+  meta = {
+    description = "Bittensor wallet implementation";
+    homepage = "https://github.com/latent-to/btwallet";
+    changelog = "https://github.com/latent-to/btwallet/releases/tag/${finalAttrs.src.tag}";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ kilyanni ];
+  };
+})

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -2110,6 +2110,8 @@ self: super: with self; {
 
   bitstruct = callPackage ../development/python-modules/bitstruct { };
 
+  bittensor-wallet = callPackage ../development/python-modules/bittensor-wallet { };
+
   bitvavo-aio = callPackage ../development/python-modules/bitvavo-aio { };
 
   bitvector-for-humans = callPackage ../development/python-modules/bitvector-for-humans { };

